File: 2022/18.hs
{-# OPTIONS_GHC -Wno-tabs #-}
import Data.List
import Data.List.Split (splitOn)
import qualified Data.Set as S
import qualified Data.Sequence as Q
import Control.Exception

type Point = (Integer, Integer, Integer)
type PointCloud = S.Set Point

getInput :: IO PointCloud
getInput = S.fromList <$> map parsePoint <$> lines <$> readFile "18.txt"

parsePoint :: String -> Point
parsePoint s = (read x, read y, read z)
	where [x, y, z] = splitOn "," s

neighbours :: Point -> [Point]
neighbours (x, y, z) = [(x + dx, y + dy, z + dz) | (dx, dy, dz) <- [(-1,0,0), (1,0,0), (0,-1,0), (0,1,0), (0,0,-1), (0,0,1)]]

surfaceArea :: (Point -> Bool) -> PointCloud -> Integer
surfaceArea pred points = genericLength [() | p <- S.elems points, n <- neighbours p, not $ n `S.member` points, pred n]

floodFillExterior :: PointCloud -> PointCloud
floodFillExterior points = worker (Q.singleton (minx, miny, minz)) S.empty
	where
		minx = (minimum $ map (\(x,y,z) -> x) $ S.elems points) - 1
		maxx = (maximum $ map (\(x,y,z) -> x) $ S.elems points) + 1
		miny = (minimum $ map (\(x,y,z) -> y) $ S.elems points) - 1
		maxy = (maximum $ map (\(x,y,z) -> y) $ S.elems points) + 1
		minz = (minimum $ map (\(x,y,z) -> z) $ S.elems points) - 1
		maxz = (maximum $ map (\(x,y,z) -> z) $ S.elems points) + 1
		worker :: Q.Seq Point -> PointCloud -> PointCloud
		worker Q.Empty exterior = exterior
		worker (next@(x,y,z) Q.:<| todo) exterior
			| x < minx || x > maxx || y < miny || y > maxy || z < minz || z > maxz = worker todo exterior
			| next `S.member` exterior = worker todo exterior
			| next `S.member` points = worker todo exterior
			| otherwise = worker (todo Q.>< Q.fromList (neighbours next)) (S.insert next exterior)

tests :: IO ()
tests = do
	check $ surfaceArea (const True) testData == 64
	check $ surfaceArea (`S.member` exterior) testData == 58
	where
		testData = S.fromList $ map parsePoint ["2,2,2","1,2,2","3,2,2","2,1,2","2,3,2","2,2,1","2,2,3","2,2,4","2,2,6","1,2,5","3,2,5","2,1,5","2,3,5"]
		exterior = floodFillExterior testData
		check True = return ()
		check False = throwIO $ AssertionFailed "test failed"

main :: IO ()
main = do
	tests
	dat <- getInput
	let exterior = floodFillExterior dat
	print $ surfaceArea (const True) dat
	print $ surfaceArea (`S.member` exterior) dat

File: 2022/18.py
#!/usr/bin/python3
dat = {
	(4,9,7),(11,4,7),(8,18,14),[...snip...]
}

n = 0
for x,y,z in dat:
	for dx,dy,dz in [(-1,0,0),(1,0,0),(0,-1,0),(0,1,0),(0,0,-1),(0,0,1)]:
		if (x+dx,y+dy,z+dz) not in dat:
			n += 1
print(n)

minx = min(i[0] for i in dat) - 1
maxx = max(i[0] for i in dat) + 2
miny = min(i[1] for i in dat) - 1
maxy = max(i[1] for i in dat) + 2
minz = min(i[2] for i in dat) - 1
maxz = max(i[2] for i in dat) + 2

water = set()
tofill = [(minx, miny, minz)]
while tofill:
	wx, wy, wz = tofill.pop()
	if (wx, wy, wz) in water or (wx, wy, wz) in dat:
		continue
	water.add((wx, wy, wz))
	for dx,dy,dz in [(-1,0,0),(1,0,0),(0,-1,0),(0,1,0),(0,0,-1),(0,0,1)]:
		tx, ty, tz = wx+dx,wy+dy,wz+dz
		if minx <= tx < maxx and miny <= ty < maxy and minz <= tz < maxz:
			tofill.append((tx, ty, tz))

n = 0
for x,y,z in dat:
	for dx,dy,dz in [(-1,0,0),(1,0,0),(0,-1,0),(0,1,0),(0,0,-1),(0,0,1)]:
		if (x+dx,y+dy,z+dz) in water:
			n += 1
print(n)

